File: server/bilibili/type.go
package bilibili

import "encoding/json"

// BaseRes 来自 Bilibili 的接口响应，Message 字段为 msg
type BaseRes struct {
	Code    int             `json:"code"`
	Message string          `json:"msg"`
	Data    json.RawMessage `json:"data"`
}

// BaseRes2 来自 Bilibili 的接口响应，Message 字段为 message 而不是 msg
type BaseResV2 struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data"`
}

// BaseResV3 来自 Bilibili 的接口响应，Message 字段为 message 而不是 msg，Data 字段为 result
type BaseResV3 struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Result  json.RawMessage `json:"result"`
}

func (b *BaseRes) Success() bool {
	return b.Code == 0
}

func (b *BaseResV2) Success() bool {
	return b.Code == 0
}

func (b *BaseResV3) Success() bool {
	return b.Code == 0
}

type QRInfo struct {
	URL       string `json:"url"`
	QrcodeKey string `json:"qrcode_key"`
}

type QRStatus struct {
	URL          string `json:"string"`
	RefreshToken string `json:"refresh_token"`
	Code         int    `json:"code"`
	Message      string `json:"message"`
}

const (
	QR_NO_SCAN  = 86101 // 未扫码
	QR_NO_CLICK = 86090 // 已扫码未确认
	QR_EXPIRES  = 86038 // 已过期
	QR_SUCCESS  = 0     // 已确认登录
)

// Dimension 视频分辨率
type Dimension struct {
	Width  int `json:"width"`
	Height int `json:"height"`
	Rotate int `json:"rotate"`
}

// StaffItem 视频人员
type StaffItem struct {
	Mid   int    `json:"mid"`
	Title string `json:"title"`
	Name  string `json:"name"`
	Face  string `json:"face"`
}

// Page 分集信息
type Page struct {
	// 配合 Bvid 用于获取播放地址
	Cid int `json:"cid"`
	// 当前分集在合集中的序号，从 1 开始
	Page int `json:"page"`
	// 分集标题
	Part string `json:"part"`
	// 分集时长
	Duration int `json:"duration"`
	// 分集分辨率
	Dimension `json:"dimension"`
}

// 通过 BVID 获取的视频信息
type VideoInfo struct {
	Bvid    string `json:"bvid"`
	Aid     int    `json:"aid"`
	Pic     string `json:"pic"`
	Title   string `json:"title"`
	Pubdate int    `json:"pubdate"`
	Desc    string `json:"desc"`
	Owner   struct {
		Mid  int    `json:"mid"`
		Name string `json:"name"`
		Face string `json:"face"`
	} `json:"owner"`
	Dimension `json:"dimension"`
	Staff     []StaffItem `json:"staff"`
	Pages     []Page      `json:"pages"`
	Duration  int         `json:"duration"`
	Stat      struct {
		Aid      int `json:"aid"`
		View     int `json:"view"`     // 播放数量
		Danmaku  int `json:"danmaku"`  // 弹幕数量
		Reply    int `json:"reply"`    // 评论数量
		Favorite int `json:"favorite"` // 收藏数量
		Coin     int `json:"coin"`     // 投币数量
		Share    int `json:"share"`    // 转发数量
		NowRank  int `json:"now_rank"` // 当前排名
		HisRank  int `json:"his_rank"` // 历史最高排名
		Like     int `json:"like"`     // 点赞数量
		Dislike  int `json:"dislike"`  // 不喜欢
	} `json:"stat"`
}

// Episode 剧集分集信息
type Episode struct {
	Aid       int                `json:"aid"`
	Bvid      string             `json:"bvid"`
	Cid       int                `json:"cid"`
	Cover     string             `json:"cover"` // 封面
	Dimension `json:"dimension"` // 分辨率
	Duration  int                `json:"duration"` // 时长
	EPID      int                `json:"ep_id"`
	LongTitle string             `json:"long_title"` // 分集完整标题，比如【法外狂徒张三现身！】
	PubTime   int                `json:"pub_time"`   // 发布时间
	Title     string             `json:"title"`      // 分集简略标题，比如【1】
}

// 通过 EPID 获取的视频信息
type SeasonInfo struct {
	Actors string `json:"actors"` // 演员名单
	Areas  []struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	} `json:"areas"` // 地区列表
	Cover    string `json:"cover"`    // 封面
	Evaluate string `json:"evaluate"` // 简介
	Publish  struct {
		IsFinish int    `json:"is_finish"` // 是否完结
		PubTime  string `json:"pub_time"`  // 发布时间
	} `json:"publish"`
	SeasonID    int    `json:"season_id"`    // 剧集编号
	SeasonTitle string `json:"season_title"` // 剧集标题
	Stat        struct {
		Coins     int `json:"coins"`     // 投币数量
		Danmakus  int `json:"danmakus"`  // 弹幕数量
		Favorite  int `json:"favorite"`  // 收藏数量
		Favorites int `json:"favorites"` // 追剧数量
		Likes     int `json:"likes"`     // 点赞数量
		Reply     int `json:"reply"`     // 评论数量
		Share     int `json:"share"`     // 分享数量
		Views     int `json:"views"`     // 播放数量
	} `json:"stat"`
	Styles []string `json:"styles"` // 剧集内容类型，例如 [ "短剧", "奇幻", "搞笑" ]
	Title  string   `json:"title"`  // 剧集标题
	Total  int      `json:"total"`  // 总集数

	Episodes []Episode `json:"episodes"` // 分集信息列表

	NewEp struct {
		Desc  string `json:"desc"`   // 更新状态文本
		IsNew int    `json:"is_new"` // 是否是连载，0 为完结，1 为连载
	} `json:"new_ep"`
}

type PlayInfo struct {
	AcceptDescription []string `json:"accept_description"`
	AcceptQuality     []int    `json:"accept_quality"`
	SupportFormats    []struct {
		Quality        int      `json:"quality"`
		Format         string   `json:"format"`
		NewDescription string   `json:"new_description"`
		Codecs         []string `json:"codecs"`
	} `json:"support_formats"`
	Dash struct {
		Duration int     `json:"duration"`
		Video    []Media `json:"video"`
		Audio    []Media `json:"audio"`
	} `json:"dash"`
}

type Media struct {
	ID        int      `json:"id"`
	BaseURL   string   `json:"baseUrl"`
	BackupURL []string `json:"backupUrl"`
	Bandwidth int      `json:"bandwidth"`
	MimeType  string   `json:"mimeType"`
	Codecs    string   `json:"codecs"`
	Width     int      `json:"width"`
	Height    int      `json:"height"`
	FrameRate string   `json:"frameRate"`
	Codecid   int      `json:"codecid"`
}

File: client/src/work/type.ts
import { State } from 'vanjs-core'

/**
 * 参考：[qn 视频清晰度标识](https://socialsisteryi.github.io/bilibili-API-collect/docs/video/videostream_url.html#qn%E8%A7%86%E9%A2%91%E6%B8%85%E6%99%B0%E5%BA%A6%E6%A0%87%E8%AF%86)
 * 
 * | 值  | 含义        |
 * | --- | ----------: |
 * | 6   | 240P 极速   |
 * | 16  | 360P 流畅   |
 * | 32  | 480P 清晰   |
 * | 64  | 720P 高清   |
 * | 74  | 720P60 高帧率 |
 * | 80  | 1080P 高清  |
 * | 112 | 1080P+ 高码率 |
 * | 116 | 1080P60 高帧率 |
 * | 120 | 4K 超清     |
 * | 125 | HDR 真彩色  |
 * | 126 | 杜比视界     |
 * | 127 | 8K 超高清    |
 */
export type VideoFormat = 6 | 16 | 32 | 64 | 74 | 80 | 112 | 116 | 120 | 125 | 126 | 127

/** 视频解析结果，数据用于 DOM 渲染，同时兼容 BV、EP、SS */
export type VideoParseResult = {
    /** 合集标题 */
    title: string
    /** 合集描述 */
    description: string
    /** 合集发布时间 */
    publishData: string
    /** 合集封面 */
    cover: string
    /** 合集总时长 */
    duration: number
    /** 分集列表 */
    pages: PageInParseResult[]

    section: SectionItem[]

    /** 合集作者 */
    owner: {
        mid: number
        name: string
        face: string
    }

    /** 合集分辨率 */
    dimension: {
        width: number
        height: number
        rotate: number
    }
    staff: string[]
    /** 更新状态 */
    status: string
    /** 地区信息 */
    areas: string[]
    /** 分类标签 */
    styles: string[]
    /** 播放页面 */
    targetURL: string
}

export type SectionItem = {
    title: string
    pages: PageInParseResult[]
}

export type PageInParseResult = {
    /** 分集 CID */
    cid: number
    /** 分集 BVID */
    bvid: string
    /** 分集在合集中的序号，从 1 开始 */
    page: number
    /** 分集标题 */
    part: string
    /** 分集时长 */
    duration: number
    /** 分集分辨率 */
    dimension: {
        width: number
        height: number
        rotate: number
    }
    /** 前置胶囊标签内容 */
    bandge: string
    /** 是否选中 */
    selected: State<boolean>
}

export type StaffItem = {
    mid: number
    title: string
    name: string
    face: string
}

type PageInVideoInfo = {
    cid: number
    page: number
    from: string
    part: string
    duration: number
    dimension: {
        width: number
        height: number
        rotate: number
    }
}

/** 接口返回的视频信息 */
export type VideoInfo = {
    aid: number
    staff: null | StaffItem[]
    title: string
    desc: string
    pubdate: number
    pic: string
    duration: number
    bvid: string
    pages: PageInVideoInfo[]
    owner: {
        mid: number
        name: string
        face: string
    }
    dimension: {
        width: number
        height: number
        rotate: number
    },
    ugc_season: {
        sections: {
            title: string
            episodes: {
                title: string
                pages: PageInVideoInfo[]
                bvid: string
            }[]
        }[] | null
        title: string
    }
}

/** 接口返回的剧集信息 */
export type SeasonInfo = {
    actors: string
    areas: {
        id: number
        name: string
    }[]
    cover: string
    evaluate: string
    publish: {
        is_finish: number
        pub_time: string
    };
    season_id: number
    season_title: string
    stat: {
        coins: number
        danmakus: number
        favorite: number
        favorites: number
        likes: number
        reply: number
        share: number
        views: number
    }
    styles: string[]
    title: string
    total: number
    episodes: Episode[]
    new_ep: {
        desc: string
        is_new: number
    }
    section: {
        title: string
        episodes: Episode[]
    }[] | null
}

export type Episode = {
    aid: number
    bvid: string
    cid: number
    cover: string
    dimension: {
        width: number
        height: number
        rotate: number
    }
    duration: number
    ep_id: number
    long_title: string
    pub_time: number
    title: string
}

export type VideoInfoCardMode = State<"video" | "season" | "hide">

export type PlayInfo = {
    accept_description: string[]
    accept_quality: number[]
    support_formats: {
        quality: number
        format: string
        new_description: string
        codecs: string[]
    }[]
    dash: {
        duration: number
        video: {
            id: number
            baseUrl: string
            backupUrl: string[]
            bandwidth: number
            mimeType: string
            codecs: string
            width: number
            height: number
            frameRate: string
            codecid: number
        }[]
        audio: {
            id: number
            baseUrl: string
            backupUrl: string[]
            bandwidth: number
            mimeType: string
            codecs: string
            width: number
            height: number
            frameRate: string
            codecid: number
        }[]
    }
}

/** 创建任务时的初始数据 */
export type TaskInitData = {
    bvid: string
    cid: number
    format: number
    title: string
    owner: string
    cover: string
}

/** 任务数据库中的数据 */
export type TaskInDB = TaskInitData & {
    id: number
    folder: string
    createAt: string
    status: TaskStatus
}

export type TaskStatus = 'done' | 'waiting' | 'running' | 'error'